### ISSUE TITLE
[Question] Using ITestOutputHelper to capture log output

### ISSUE THREAD

**Issue by ashmind** [xunit/xunit#395]
Is there a recommended way to capture log output (log4net, LibLog+Serilog, etc) to `ITestOutputHelper`, when static logger is used by tested classes? E.g. there is quite a bit of code using `LogProvider.GetCurrentClassLogger` or `LogManager.GetLogger`.

To be fair I'm normally injecting the logger -- but is it the only way going forward? Or is there some simple solution for `ITestOutputHelper`-based appender (considering the parallelism)?

**Comment by bradwilson** [comment]
This is the code we used in v1 to capture everything:

https://github.com/xunit/xunit/blob/v1/src/xunit/Sdk/Commands/TestCommands/ExceptionAndOutputCaptureCommand.cs

Just be aware that the reasons we don't do this any more -- because these are shared static resources and parallel tests could be accessing them -- still apply.

Here's one strategy you could use: a collection fixture that does the capture and reporting work. It solves the parallelization problem (since every class that wanted that fixture would be put into the test collection). So, something roughly like this:

```csharp
// Fixture which does the work
public class CaptureFixture
{
    public void Capture()
    {
        // ... capture whatever outputs you want captured, to a buffer
    }

    public void Report(ITestOutputHelper output)
    {
        output.WriteLine(/* buffered output */);
        // ... release whatever was captured
    }
}

// Collection definition
[CollectionDefinition("Capture")]
public class CaptureTestCollection : ICollectionFixture<CaptureFixture> { }

// Base class for test classes that require capturing
[Collection("Capture")]
public abstract class CaptureTests : IDisposable
{
    CaptureFixture fixture
    ITestOutputHelper output;

    protected CaptureTests(CaptureFixture fixture, ITestOutputHelper output)
    {
        this.fixture = fixture;
        this.output = output;

        fixture.Capture();
    }

    public void Dispose()
    {
        fixture.Report(output);
    }
}

public class MyTestsWhichNeedCapture : CaptureTests
{
    public MyTestsWhichNeedCapture(CaptureFixture fixture, ITestOutputHelper output)
        : base(fixture, output) { }

    // ... tests ...
}
```

Using the base class approach is optional, but it removes all the boilerplate code that you'd need for every test. You just need to add the delegating constructor.

For more information on fixtures, see: http://xunit.github.io/docs/shared-context.html

**Comment by bradwilson** [comment]
This is the capture part:
https://github.com/xunit/xunit/blob/9aef27cd20ce6bb06287dc37c490f50b69bdc55d/src/xunit/Sdk/Commands/TestCommands/ExceptionAndOutputCaptureCommand.cs#L36-L56

And this is the cleanup part:
https://github.com/xunit/xunit/blob/9aef27cd20ce6bb06287dc37c490f50b69bdc55d/src/xunit/Sdk/Commands/TestCommands/ExceptionAndOutputCaptureCommand.cs#L61-L74

**Comment by ashmind** [comment]
Thanks! But that's essentially removing parallelism, isn't it?
I wonder if I can do something else instead, e.g. use `CallContext` -- I'll try later when I have time.

**Comment by bradwilson** [comment]
What you do to make parallelism work is: stop using shared resources. :)

**Comment by ashmind** [comment]
> What you do to make parallelism work is: stop using shared resources. :)
I do actually agree with that -- but in practice loggers (e.g. log4net) are actually thread-safe and I see a lot of ppl using static loggers (especially since all examples promote it).

**Comment by bradwilson** [comment]
You're welcome to put something in `CallContext` that can help you track down the original caller. We can't do that in the framework itself, because we run in places where `CallContext` is not available (and there is no suitable Task-aware replacement).

**Comment by ejsmith** [comment]
This is a really big PITA. I understand the reasoning and it is certainly logical to default to not including output since it could be very confusing to have the test output mixed up with other tests running at the same time. But you have to understand that it is extremely common to use unit tests to debug code. In a lot of cases you can just step through and see what is going on, but in cases where the issue is time sensitive that isn't possible. Could you not provide an option to opt into getting output and make it clear what could happen as the result of that choice?

**Comment by ejsmith** [comment]
This is what I ended up doing.

https://github.com/exceptionless/Foundatio/blob/master/src/Core/Tests/Utility/CaptureFixture.cs

**Comment by damianh** [comment]
Blogged about it how I'm doing it here http://dhickey.ie/2015/06/capturing-log-output-in-tests-with-xunit2/

tl;dr use diagnostic contexts to correlate the log messages and filter + pipe them to the correct ITestOutputHelper instance.

**Comment by bradwilson** [comment]
Given there are multiple workarounds, we've decided not to do this feature.

**Comment by urbanhop** [comment]
Unfortunately no workaround exists for F# users: The nice thing of xunit is to allow plain functions inside modules as tests. It is however not possible to capture ITestOutput in this scenario.

**Comment by bradwilson** [comment]
Why don't you think using a class in F# is an acceptable workaround?

**Comment by urbanhop** [comment]
Well, if classes for tests would be fine for me, I would switch to NUnit and have console output and be fine. The reason why F# devs prefer xunit over NUnit is that plain functions are so much simpler to handle. It starts by being less typing and easier reading:

    module Tests

    open Xunit

    [<Fact()>]
    let testfun1() =
        ()

    [<Fact()>]
    let testfun2() =
        ()

versus with classes

    module Test.ACME

    open NUnit.Framework

    type MyTests() =

      [<TestCase>]
      member x.test1() =
        ()

      [<TestCase>]
      member x.test2() =
        ()

The much greater benefit however is, that functions are the unit of work in funtional programming. member functions are more the exception. And this again matters, because as an F# programmer I execute code during development mainly as script code in the REPL and for that purpose I highlight relevant functions and send them to the REPL. With members this is not possible, so with classes and member methods as tests, the workflow is broken and I have to add and remove "member o." and indent code all the time. Writing tests in F# means moving code between script and test all the time. Also, when debugging tests, you cope the test code to the script and work there. This is so bad, that I much prefer sending output to the debug stream and view it with DebugViewer.

Sometime however, one would just add console output to the test instead of copying over to a script and this is where I am stuck with xunit.

The F# test functions get compiled into members of their module, which becomes a CLR class. Since an F# module cannot implement an interface, I found no way to work around. I also considered attributes, with no outcome. Finally I decided to use Trace.WriteLine, but would certainly prefer console output in the test runner.

I tried already to create my own xunit that allows me to write to the console output, but found that the implementation is actually in the runners that host the tests. Is that correct?

**Comment by bradwilson** [comment]
I appreciate that it might not be a workaround you like, but that's different from your original assertion that "there are no workarounds".

**Comment by urbanhop** [comment]
Agreed ;-)

Am I right, that changing this, would require change of the runners ?

**Comment by bradwilson** [comment]
You would need to write a custom implementation of `ITestFramework`, because the way the built in test framework works for things like this is constructor injection. You would need to find some equivalent (which, by the way, I don't necessarily think exists).

**Comment by bradwilson** [comment]
The TestClassRunner is where that work is done today:

https://github.com/xunit/xunit/blob/b1cbc40158cc53be5b9150bcdeb2e1cae40db5c9/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestClassRunner.cs#L163-L167

**Comment by urbanhop** [comment]
Very helpful, thanks. 

I would do some rude private hack, that allows Console to be temporarily turned on during test design - via configuration or just by copying hacked xunit assemblies. Maybe trace output is good enough though.

**Comment by Michael-DST** [comment]
Sorry to resurrect this thread, but this is the closest thing I have seen online towards what I am looking to do.  I get the idea of a ITestOutputHelper, but how do we pipe the results from say Debug.WriteLine to this interface?  I do have to thank @damianh for his blog post.  That is close to what I am looking for, but for a simpler imlpementation (namely, no external loggers, just Debug/Trace/Console -> WriteLine.

In my case, I am experiencing tests failing that only occur during Release (non-Debug) mode in parallel, so I cannot simply debug this issue in the debugger.

**Comment by Michael-DST** [comment]
OK not only did I resurrect this thread, I did so with a poor understanding of the problem at hand, apologies.  It seems that Debug/Trace/Console are shared resources and it sounds like they are very difficult to use in a parallel environment.  This is disappointing as much as frustrating, obviously, as logging seems like such a simple/obvious task/scenario it would be baked into any unit testing framework like the pass/fail symbols are.  That's no knock on xUnit -- everyone does a great job here and I enjoy the framework very much.  I have been studying the code and it is tops (of course).  There is definitely lots of awesomeness here and I am nothing short of impressed with what I have seen so far.  This would be the first "disappointment" that I have encountered, for lack of a better word.

That is, it just seems like logging is an obvious scenario (to me, at least) and one that would be plainly accounted for from the start that I am surprised that I am in the minority here in my expectations with this.

In any case, I am at the very least concerned that .NET TPL is designed in such a way that getting simple logging messages from a thread is a near (or out-right) impossible ordeal.  That is the bigger worry for me with all things considered.  Hopefully, the TPL team is aware of this shortcoming and this will be addressed in a future release?  It seems like being able to pair an ID from the current thread/executing task with a text message would be a simple and obvious implementation, but I have not looked much into this and feel I have already spent to much time on a feature that again I feel should be available from the outset and already present.

I will quit my quasi-semi-rant while I am ahead... I think/hope. :)

**Comment by bradwilson** [comment]
The issue isn't really one of logging, but one of mutable static state (in this case, the implementation of Console) not being compatible with parallelism.

Additionally, while we did capture Console in v1, it was a poor design decision, because it reused a system component in a way in which it wasn't intended. We got bugs from people trying to test the trace handlers about how our interception of Trace caused them significant issues, so we decided that we would start with our own interface, and let people decide how best to adapt that to whatever logging systems they may already be using.

**Comment by Michael-DST** [comment]
OK cool @bradwilson thank you for the additional context.  I am starting to wrap my head around this a little more and think I have a solution that won't involve too much effort after all.  

As an aside, I am appreciative of xUnit 2.0 baking parallelism right into its foundation, as it has not only made me investigate TPL further, but has exposed some incompatible code in my own framework that is also suffering from the  problem/reason that you outline above.  However, I have the luxury of refactoring/adapting to these requirements whereas core/legacy .NET components do not. :)

As awesome as TPL is, it still is version 1 (or 2?) and it is these sorts of challenges ([and others](http://visualstudio.uservoice.com/forums/121579-visual-studio/suggestions/9126493-improve-asynchronous-programming-model))  that really underscore the amount of work that remains to close the gap between sync/async programming.

Chalk this up to growing pains, I guess!
